openedx/core/djangoapps/course_groups/models.py

```diff
@@ -95,7 +95,6 @@ def save(self, *args, **kwargs):
         if 'force_insert' in kwargs and kwargs['force_insert'] is True:
             with transaction.atomic():
                 self.course_user_group.users.add(self.user)
-                self.course_user_group.save()
```

Comment:
As discussed offline, removing this call *might* fix the deadlocking issue, as this creates a scenario where multiple processes might be trying to atomically update multiple databases in a single transaction.

```
mysql> show columns in course_groups_courseusergroup_users;
+--------------------+---------+------+-----+---------+----------------+
| Field              | Type    | Null | Key | Default | Extra          |
+--------------------+---------+------+-----+---------+----------------+
| id                 | int(11) | NO   | PRI | NULL    | auto_increment |
| courseusergroup_id | int(11) | NO   | MUL | NULL    |                |
| user_id            | int(11) | NO   | MUL | NULL    |                |
+--------------------+---------+------+-----+---------+----------------+
3 rows in set (0.00 sec)

mysql> show columns in course_groups_courseusergroup;
+------------+--------------+------+-----+---------+----------------+
| Field      | Type         | Null | Key | Default | Extra          |
+------------+--------------+------+-----+---------+----------------+
| id         | int(11)      | NO   | PRI | NULL    | auto_increment |
| name       | varchar(255) | NO   | MUL | NULL    |                |
| course_id  | varchar(255) | NO   | MUL | NULL    |                |
| group_type | varchar(20)  | NO   |     | NULL    |                |
+------------+--------------+------+-----+---------+----------------+
4 rows in set (0.00 sec)
```